### DISCUSSION TITLE
Websocket for streaming endpoint unable to buffer bytes correctly

### DISCUSSION THREAD

**Discussion by asusdisciple** [fastapi/fastapi#12906] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Server side
@streaming_router.websocket("/audio-stream/")
async def audio_stream_endpoint(websocket: WebSocket):

    buffer = bytearray()
    buffer_size = 1024
    await websocket.accept()
    try:
        while True:
            # Receive raw audio data from the client
            data = await websocket.receive_bytes()
            buffer += data
            if len(buffer) >= buffer_size:
                chunk = buffer[:buffer_size]
                buffer = buffer[buffer_size:]  
                await websocket.send_bytes(chunk)
```

### Description

So I have the problem that I want to send a audio stream to the server, process it and send it back to the client. This worked
fine as long as I did not use any buffers and just called `send_bytes()` right after `receive_bytes()`. However as soon as I integrate
very simple buffer logic like above the audio signals crumbles. Its still there but with very high latency and a lot of artifacts.

I tried to use a BytesIO object as buffer but this did not work at all. I think I am missing something really basic here, but I cant figure this out alone and I am at my wits end. Maybe reading and writing from the same buffer object is the problem?

I just though the FastApi async methodology would handle stuff like this automatically.

### Operating System

Linux

### Operating System Details

Ubuntu 24.04
Python 3.10

### FastAPI Version

0.115.2

### Pydantic Version

2.9.2

### Python Version

3.10

**Comment by YuriiMotov** [comment]
Try changing `if len(buffer) >= buffer_size:` to `while len(buffer) >= buffer_size:`.

If data comes to your websocket in bigger chunks (e.g. 4096 byte), it will send only 1024 and leave other 3072 in the buffer. Then it receives another 4096 and send 1024, buffer is already 6144.
So, you need to send not just one chunk, but repeat it untill buffer size is less than chunk size.

**Comment by asusdisciple** [comment]
Unfortunately that does not really change anything. Also The stream signal which I send back seems to get "slower" in the sense that the voice gets a Slomo effect, which gets stronger with the duration of the stream I send.

This is really strange behaviour.

**Comment by luzzodev** [comment]
While @YuriiMotov has highlighted one of the issues concerning the buffer, the problem of real-time audio streaming is much more complex.

while it's true that you should iterate over your buffer send data back, in the solution proposed will still remain `nbytes < len(buffer) ` of unseded bytes after the last received message.

Said that this approach if only valid if you are working with no-realtime-audio and in this case, for testing purpose only, you can check that 

> `all_bytes_sent_from_client == all_bytes_received_from_client`

Of course while testing do not alter or transfrom bytes on server side.

For realtime-audio-striming thinks are mutch more complex:
    Assuming you are working with raw audio and not encoded one (thinks are even more comples here)
- Playing an audio require a specific rate.
- The audio rate is based on time.
- Network latency or processing overhead matters a lot.

Keeping in mind that using websocket could not be the best protocol to use, here you have some advice:

- Using a buffer server side and sending a fixed lenght will not work as expected.
  - you are forcing sending data based on lenght and not on time/rate.
  - you must always send data according to the audio-rate

Considering you mentioned that audio seems like "slowed" i would assume your rate is lower then expected. ( this is one of the reasons why without buffer works and with it don''t).

Even without buffer this could not works but it depends on the network latency and on the bytes sent from client.

1. Sending too few bytes could not be "fixed" in any way then increase them.
2. Sending too many bytes could be fixed with a buffer client-side from where the player should get the bytes according to the audio rate. This is a good approach even to mitigate network latency or processing delay.
